src/Components/Web.JS/src/Rendering/JSRootComponents.ts

```diff
@@ -7,100 +7,170 @@ let nextPendingDynamicRootComponentIdentifier = 0;
-    }
+  const container = pendingRootComponentContainers.get(containerIdentifier);
+  if (container) {
+    pendingRootComponentContainers.delete(containerIdentifier);
+    return container;
+  }
+}
+
+class EventCallbackWrapper {
+  private _wrapper: { func: any };
+
+  private readonly _jsObjectReference: any;
+
+  constructor(eventCallback: any) {
+    this._wrapper = { func: eventCallback };
+    this._jsObjectReference = DotNet.createJSObjectReference(this._wrapper);
+  }
+
+  setCallback(eventCallback: any) {
+    this._wrapper.func = eventCallback;
+  }
+
+  getJSObjectReference() {
+    return this._jsObjectReference;
+  }
+
+  dispose() {
+    DotNet.disposeJSObjectReference(this._jsObjectReference);
+  }
 }
 
 class DynamicRootComponent {
-    private _componentId: number | null;
+  private _componentId: number | null;
 
-    constructor(componentId: number) {
-        this._componentId = componentId;
-    }
+  private _jsEventCallbackWrappers = new Map<string, EventCallbackWrapper | null>();
+
+  constructor(componentId: number, parameters: JSComponentParameter[]) {
+    this._componentId = componentId;
 
-    setParameters(parameters: ComponentParameters) {
-        parameters = parameters || {};
-        const parameterCount = Object.keys(parameters).length;
-        return getRequiredManager().invokeMethodAsync('SetRootComponentParameters', this._componentId, parameterCount, parameters);
+    for (const parameter of parameters) {
+      if (parameter.type === 'eventcallback') {
+        // We lazily initialize event callback wrappers, so null entries are added for now.
+        this._jsEventCallbackWrappers.set(parameter.name.toLowerCase(), null);
+      }
+    }
+  }
+
+  setParameters(parameters: ComponentParameters) {
+    parameters = parameters || {};
+
+    const keys = Object.keys(parameters);
+    const parameterCount = keys.length;
+
+    // We iterate over the provided parameters to modify event callback values.
+    for (const key of keys) {
+      const keyLowerCase = key.toLowerCase();
+      let callbackWrapper = this._jsEventCallbackWrappers.get(keyLowerCase);
+      if (callbackWrapper === undefined) {
+        // Not an event callback.
+        continue;
+      }
+
+      const callback = parameters[key];
+      if (callbackWrapper === null) {
+        // This is the first time this parameter has been provided.
+        // Create the wrapper, then replace the existing parameter value with the JS object reference.
+        callbackWrapper = new EventCallbackWrapper(callback);
+        this._jsEventCallbackWrappers.set(keyLowerCase, callbackWrapper);
+        parameters[key] = callbackWrapper.getJSObjectReference();
+      } else {
+        // This parameter has been seen before. Since the existing JS object reference is still valid in .NET,
+        // we can update the callback on the JS side and skip this parameter.
+        callbackWrapper.setCallback(callback);
+        delete parameters[key];
```

Comment:
I've since realized this is not the case - the function is cached internally so changing it here won't actually make a difference. We need a new JSObjectReference altogether.

Reply:
I suspect we could leave the same JSObjectReference in place, but call the callback through a layer of indirection, e.g.:

```ts
class EventCallbackWrapper {
    invokeCallback(arg: any): Task {
        return this._func();
    }
}
```

Then we can swap out `_func` without changing the function instance that JS interop is trying to invoke directly.

Reply:
Also just want to check if there should also be a third case here. What if someone explicitly sets an eventcallback parameter to `null` after it was previously nonnull? I'd expect the .NET side to receive a `null` so that it can overwrite its local `EventCallback` with a `default` value (so that `eventCallbackInstance.HasDelegate` would become false). Looks like the logic as-is wouldn't notify .NET, and it would continue thinking there was a delegate.

Reply:
Finally, I also wonder if we should definitely be mutating the `parameters` object that the developer passes in here. If we do that, then the caller can observe things like the replacement of their function values with `EventCallbackWrapper` instances, and can even do weird things like call `dispose` on them or otherwise interfere with the internals of our system (since the caller still has access to the `parameters` object that they passed in).

As an alternative, we could consider replacing the `_jsEventCallbackWrappers` map with some similar map that includes all the most recently-supplied parameters (not just eventcallback ones), and then we'd be looping over all the supplied parameters, copying them into the new map, with a bit of special-case logic for eventcallback ones that need to be instantiated or updated if they already exist in the map.

Of course if we do that it would invalidate my suggestion [here](https://github.com/dotnet/aspnetcore/pull/35206/files?w=1#r686638284). I don't personally mind which way you choose to go on this. Just raising the question - depends on whether you think it's an issue that the caller could mutate/observe `parameters` even further after they called us.

Reply:
If we had a map that included the most recently-supplied parameters, passing that to .NET instead of the user-provided `parameters` object, wouldn't all parameters be re-supplied to .NET each time rather than only the changed parameters? Or am I missing something?

I've modified my implementation to generate a copy of the `parameters` object in each call to `setParameters`, with special handling for properties whose names appear in the `_jsEventCallbackWrappers` map. It also now properly handles supplying `null` eventcallback values. Let me know if you see any issues with the way I've done things - happy to make a 2nd revision 😃